modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/handlers/BuildIndexCommandHandler.java

```diff
@@ -0,0 +1,185 @@
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.partition.replicator.network.command.BuildIndexCommand;
+import org.apache.ignite.internal.partition.replicator.raft.handlers.AbstractCommandHandler;
+import org.apache.ignite.internal.partition.replicator.raft.snapshot.PartitionDataStorage;
+import org.apache.ignite.internal.schema.BinaryRow;
+import org.apache.ignite.internal.schema.BinaryRowUpgrader;
+import org.apache.ignite.internal.schema.SchemaDescriptor;
+import org.apache.ignite.internal.schema.SchemaRegistry;
+import org.apache.ignite.internal.storage.BinaryRowAndRowId;
+import org.apache.ignite.internal.storage.MvPartitionStorage.Locker;
+import org.apache.ignite.internal.storage.RowId;
+import org.apache.ignite.internal.table.distributed.StorageUpdateHandler;
+import org.apache.ignite.internal.table.distributed.index.IndexMeta;
+import org.apache.ignite.internal.table.distributed.index.IndexMetaStorage;
+import org.apache.ignite.internal.table.distributed.index.MetaIndexStatusChange;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Raft command handler that process {@link BuildIndexCommand}.
+ */
+public class BuildIndexCommandHandler extends AbstractCommandHandler<BuildIndexCommand> {
+    private static final IgniteLogger LOG = Loggers.forClass(BuildIndexCommandHandler.class);
+
+    /** Data storage to which the command will be applied. */
+    private final PartitionDataStorage storage;
+
+    private final IndexMetaStorage indexMetaStorage;
+
+    /** Handler that processes storage updates. */
+    private final StorageUpdateHandler storageUpdateHandler;
+
+    private final SchemaRegistry schemaRegistry;
+
+    /**
+     * Creates a new instance of the command handler.
+     *
+     * @param storage Partition data storage.
+     * @param indexMetaStorage Index meta storage.
+     * @param storageUpdateHandler Storage update handler.
+     * @param schemaRegistry Schema registry.
+     */
+    public BuildIndexCommandHandler(
+            PartitionDataStorage storage,
+            IndexMetaStorage indexMetaStorage,
+            StorageUpdateHandler storageUpdateHandler,
+            SchemaRegistry schemaRegistry
+    ) {
+        this.storage = storage;
+        this.indexMetaStorage = indexMetaStorage;
+        this.storageUpdateHandler = storageUpdateHandler;
+        this.schemaRegistry = schemaRegistry;
+    }
+
+    @Override
+    protected IgniteBiTuple<Serializable, Boolean> handleInternally(
+            BuildIndexCommand command,
+            long commandIndex,
+            long commandTerm,
+            @Nullable HybridTimestamp safeTimestamp
+    ) throws IgniteInternalException {
+        // Skips the write command because the storage has already executed it.
+        if (commandIndex <= storage.lastAppliedIndex()) {
+            return new IgniteBiTuple<>(null, false);
+        }
```

Comment:
Would it make sense to move this logic to a template method calling `handleInternally()`? As it seems that every handler (or most of them) do this check. Not in this PR, of course.

Reply:
Yes, I thought about it, but it doesn't seem to improve the readability of the code very much. I think we can come back to this a little later, when we implement all the handlers, for example.